homeassistant/components/zha/light.py

```diff
@@ -224,17 +228,51 @@ async def async_turn_on(self, **kwargs):
         effect = kwargs.get(light.ATTR_EFFECT)
         flash = kwargs.get(light.ATTR_FLASH)
 
-        if brightness is None and self._off_brightness is not None:
+        # If the light is currently off but a turn_on call with a color/temperature is sent,
+        # the light needs to be turned on first at a low brightness level where the light is immediately transitioned
+        # to the correct color. Afterwards, the transition is only from the low brightness to the new brightness.
+        # Otherwise, the transition is from the color the light had before being turned on to the new color.
+        # This can look especially bad with transitions longer than a second.
+        color_provided_from_off = (
+            not self._state
+            and brightness_supported(self._attr_supported_color_modes)
+            and (light.ATTR_COLOR_TEMP in kwargs or light.ATTR_HS_COLOR in kwargs)
+        )
+        final_duration = duration
+        if color_provided_from_off:
+            # Set the duration for the color changing commands to 0.
+            duration = 0
+
+        if (
+            brightness is None
+            and (self._off_with_transition or color_provided_from_off)
+            and self._off_brightness is not None
+        ):
             brightness = self._off_brightness
 
+        if brightness is not None:
+            level = min(254, brightness)
+        else:
+            level = self._brightness or 254
+
         t_log = {}
-        if (brightness is not None or transition) and brightness_supported(
-            self._attr_supported_color_modes
+
+        if color_provided_from_off:
+            # If the light is currently off, we first need to turn it on at a low brightness level with no transition.
+            # After that, we set it to the desired color/temperature with no transition.
+            result = await self._level_channel.move_to_level_with_on_off(2, 0)
```

Comment:
Can we make a constant for 2. Something like `DEFAULT_MIN_BRIGHTNESS` and maybe use DEFAULT_TRANSITION which is 1 I believe…